exporter/exporter.go

```diff
@@ -0,0 +1,166 @@
+// Package exporter defines the interfaces required to implement in
+// order to add additional exporters.
+package exporter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/krakend/krakend-otel/config"
+	"github.com/krakend/krakend-otel/exporter/otelcollector"
+	"github.com/krakend/krakend-otel/exporter/prometheus"
+)
+
+// MetricReader is the interface required in order to
+// export metrics.
+type MetricReader interface {
+	MetricReader() sdkmetric.Reader
+}
+
+// SpanExporter is the interface required in order to
+// export traces.
+type SpanExporter interface {
+	SpanExporter() sdktrace.SpanExporter
+}
+
+// ExporterFactory is the function type to obtain exporters, that can
+// implement [MetricReader], [SpanExporter] or both.
+type ExporterFactory func(context.Context, map[string]interface{}) (interface{}, error)
+
+var (
+	exporterFactories map[string]ExporterFactory
+
+	metricsInstances map[string]MetricReader
+	tracesInstances  map[string]SpanExporter
+
+	mu           = new(sync.RWMutex)
+	registerOnce = new(sync.Once)
+)
+
+// RegisterExporterFactory adds an ExporterFactory associated
+// with a name, to the global exporterFactories varialbe.
+func RegisterExporterFactory(name string, ef ExporterFactory) {
+	mu.Lock()
+	exporterFactories[name] = ef
+	mu.Unlock()
+}
+
+// RegisterKnownFactories registers all known exporter factories
+func RegisterKnownFactories() {
+	registerOnce.Do(func() {
+		mu.Lock()
+		exporterFactories = map[string]ExporterFactory{
+			prometheus.ExporterKey:    prometheus.Exporter,
+			otelcollector.ExporterKey: otelcollector.Exporter,
+		}
+		metricsInstances = make(map[string]MetricReader)
+		tracesInstances = make(map[string]SpanExporter)
+		mu.Unlock()
+	})
+}
+
+// Instances create instances for a given configuration.
+func Instances(cfg *config.Config) (map[string]MetricReader, map[string]SpanExporter, []error) {
+	m := make(map[string]MetricReader)
+	s := make(map[string]SpanExporter)
+	var errList []error
+
+	ctx := context.Background()
+	mu.RLock()
+
+	for name, ecfg := range cfg.Exporters {
+		f, ok := exporterFactories[ecfg.Kind]
+		if !ok {
+			err := fmt.Errorf("Exporter of kind: %s not found", ecfg.Kind)
+			errList = append(errList, err)
+			continue
+		}
+		i, err := f(ctx, ecfg.Config)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		if i == nil {
+			err := fmt.Errorf("Implementation of kind: %s creates nil instance", ecfg.Kind)
+			errList = append(errList, err)
+		}
+		if ss, ok := i.(SpanExporter); ok && ss != nil {
+			s[name] = ss
+		} else if mm, ok := i.(MetricReader); ok {
```

Comment:
`mm` can't be nil?

Reply:
should not happen, but since we are not sure, I will add the check as in line 91 , good catch!